### ISSUE TITLE
Rule to disallow property references in destructuring patterns

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#14822]
**Please describe what the rule should do:**

Disallows references within object or array destructuring pattern values if they are not simple variables (if they are a property reference).

Computed properties are still allowed as are default values of any kind.

**What new ECMAScript feature does this rule relate to?**

Object and array destructuring.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
var obj = {};
({ a: this.b } = obj)

var obj = {};
({ a: (true).b } = obj)

var arr = [];
([ this.a ] = arr)
```

**Why should this rule be included in ESLint (instead of a plugin)?**

I know this rule is not about ECMAScript features added within the last 12 months (perhaps not offered before because it is fairly obscure), I wanted to submit in case such instances would be similarly understood as footguns that would be worth being reported. The confusability with destructuring default values and ability to affect properties in a non-readily understood way may suggest potential problems for projects.

**Are you willing to submit a pull request to implement this rule?**

Yes. I have prepared one.

**Comment by mdjermanovic** [comment]
> ```js
> ({ a: this.b } = obj)
> ```

Can you elaborate more on why this looks like a potential error? It seems to me that this rule would disallow valid code.

**Comment by brettz9** [comment]
It does allow valid code, but the user might think they are doing:

```js
({ a = this.b } = obj)
```

It is clearer I think to see that assignment side effects are occurring with a simple:

```js
this.b = obj.a;
```

One also can avoid setting variables anywhere and everywhere in the middle of some line and instead confine destructuring to changing simple variables.

**Comment by brettz9** [comment]
Oh, and apologies, I see my commit and PR title still had an old message (just now fixed) as I was evolving in my understanding of what the property rules were for. (I originally thought there were no side effects here, but discovered, as you said too, that this is valid code that *can* serve a purpose, but just one I still think is too confusable with destructured defaults (like similar rules to require wrapping regexes in parentheses due to the visual lack of clarity) and also too distracting for setting values on complex objects--similar perhaps to the rule against inline assignments.)

**Comment by mdjermanovic** [comment]
> It does allow valid code, but the user might think they are doing:
> 
> ```js
> ({ a = this.b } = obj)
> ```
> 
> It is clearer I think to see that assignment side effects are occurring with a simple:
> 
> ```js
> this.b = obj.a;
> ```

Thanks for clarifying! However, I'm still not convinced that this rule belongs to the core. `({ a: this.b } = obj)` is valid code, so assuming that it is a possible error just because of the similarity with `({ a = this.b } = obj)` doesn't seem right. Therefore, this looks more like a stylistic rule, and I'm not sure if avoiding property references in destructuring patterns is generally considered as best practice.

**Comment by nzakas** [comment]
I'm also of the mind that this isn't a good candidate for a core rule. The potential issue is pretty obscure, and the bar for including rules in the ESLint core is very high at this point. I'm not sure this type of issue occurs frequently enough to warrant inclusion.

For context, the reason why we are limiting new rules to new syntax is because the dangerous or frail patterns of usage are typically discovered within that first year of use. After that, we tend to find more style preferences than anything else, and we don't want to take on the maintenance burden for that. That's why we encourage people to create custom rules frequently.

**Comment by brettz9** [comment]
Ok, sure. I had started with the rule thinking that form of destructuring was useless (interesting to learn it isn't), but think I may still seek to add it to another plugin. Thanks!
